InvenTree/plugin/installer.py

```diff
@@ -0,0 +1,239 @@
+        raise ValidationError(_("Permission denied: only staff users can install plugins"))
+
+    logger.debug("install_plugin: %s, %s", url, packagename)
+
+    # Check if we are running in a virtual environment
+    # For now, just log a warning
+    in_venv = sys.prefix != sys.base_prefix
+
+    if not in_venv:
+        logger.warning("InvenTree is not running in a virtual environment")
+
+    # build up the command
+    install_name = ['install', '-U']
+
+    full_pkg = ''
+
+    if url:
+        # use custom registration / VCS
+        if True in [identifier in url for identifier in ['git+https', 'hg+https', 'svn+svn', ]]:
+            # using a VCS provider
+            if packagename:
+                full_pkg = f'{packagename}@{url}'
+            else:
+                full_pkg = url
+        else:  # pragma: no cover
+            # using a custom package repositories
+            # This is only for pypa compliant directory services (all current are tested above)
+            # and not covered by tests.
+            if url:
+                install_name.append('-i')
+                full_pkg = url
+            elif packagename:
+                full_pkg = packagename
+
+    elif packagename:
+        # use pypi
+        full_pkg = packagename
+
+    install_name.append(full_pkg)
+
+    ret = {}
+
+    # Execute installation via pip
+    try:
+        result = pip_command(*install_name)
+
+        ret['result'] = ret['success'] = _("Installed plugin successfully")
+        ret['output'] = str(result, 'utf-8')
+
+        if packagename:
+            if path := check_package_path(packagename):
+                # Override result information
+                ret['result'] = _(f"Installed plugin into {path}")
+
+    except subprocess.CalledProcessError as error:
+        # If an error was thrown, we need to parse the output
+
+        output = error.output.decode('utf-8')
+        logger.exception("Plugin installation failed: %s", str(output))
+
+        errors = [
+            _("Plugin installation failed"),
+        ]
+
+        for msg in output.split("\n"):
+            msg = msg.strip()
+
+            if msg:
+                errors.append(msg)
+
+        if len(errors) > 1:
+            raise ValidationError(errors)
+        else:
+            raise ValidationError(errors[0])
+
+    # Save plugin to plugins file
+    add_plugin_to_file(full_pkg)
+
+    # Reload the plugin registry, to discover the new plugin
+    from plugin.registry import registry
+    registry.reload_plugins(full_reload=True, force_reload=True, collect=True)
+
+    return ret
+
+
+def uninstall_plugin(plugin_name):
+    """Uninstall a plugin from the python virtual environment.
+
+    Note: This is intended to be implemented in a future version
+
+    - Run pip uninstall (only if plugin is installed in venv)
+    - Reverse any migrations for plugins which implement the app mixin
+    - Raise error if located in "plugins" or external plugin dir
+    """
+    ...
```

Comment:
Is this intended to be here?

Reply:
No, was a placeholder that I missed. Thanks :)